### PR TITLE
[SHELL32] Fix 'Create New' context menu under the desktop folder

### PR DESCRIPTION
**Pull request by bho3538**
## Purpose
Access the current folder directly from your desktop and if you create a new file or folder through the context menu, item does not showed it until manually refresh it.

Before
fix_before

After
fix_after

## Proposed changes

Original codes notify new item using filesystem path. so item does not update when current folder is under 'Desktop' namespace extension.

New codes notify new item using PIDL from current parent folder. so item now update when current folder is under 'Desktop' namespace extension.

### CODE REVIEW

File: dll/win32/shell32/CNewMenu.cpp [hunk 1]
```diff
@@ -406,38 +406,51 @@ CNewMenu::SHELLNEW_ITEM *CNewMenu::FindItemFromIdOffset(UINT IdOffset)
 
 HRESULT CNewMenu::SelectNewItem(LONG wEventId, UINT uFlags, LPWSTR pszName, BOOL bRename)
 {
-    CComPtr<IShellBrowser> lpSB;
+    CComPtr<IShellFolder> lpDesktopSF;
+	CComPtr<IShellFolder> lpCurrentSF;
```
Comment (RepOfCATV95) [inline dll/win32/shell32/CNewMenu.cpp:410]: Wrong code indentation.

**Review by RepOfCATV95 — CHANGES_REQUESTED** [review]

File: dll/win32/shell32/CNewMenu.cpp (hunk 1 above)
Comment (bho3538) [reply dll/win32/shell32/CNewMenu.cpp]: I changed code indentation.

File: dll/win32/shell32/CNewMenu.cpp [hunk 2]
```diff
@@ -406,38 +406,51 @@ CNewMenu::SHELLNEW_ITEM *CNewMenu::FindItemFromIdOffset(UINT IdOffset)
...
 {
-    CComPtr<IShellBrowser> lpSB;
+    CComPtr<IShellFolder> lpDesktopSF;
+    CComPtr<IShellFolder> lpCurrentSF;
     CComPtr<IShellView> lpSV;
     HRESULT hr = E_FAIL;
-    LPITEMIDLIST pidl;
-    PITEMID_CHILD pidlNewItem;
+    LPITEMIDLIST pidlFullNewItem = NULL;
+    PIDLIST_ABSOLUTE pidlNewItem = NULL;
```
Comment (learn-more) [inline dll/win32/shell32/CNewMenu.cpp:414]: If you use an `CComHeapPtr<ITEMIDLIST_ABSOLUTE>` you do not need to free it manually

File: dll/win32/shell32/CNewMenu.cpp [hunk 3]
```diff
@@ -406,38 +406,51 @@ CNewMenu::SHELLNEW_ITEM *CNewMenu::FindItemFromIdOffset(UINT IdOffset)
...
+        if (pidlNewItem) {
+            pidlFullNewItem = ILCombine(m_pidlFolder, pidlNewItem);
+			
+            /* Notify the view object about the new item */
+            SHChangeNotify(wEventId, uFlags, (LPCVOID)pidlFullNewItem, NULL);
+            SHFree(pidlFullNewItem);
+        }
+    }
+	
+	/* Get a pointer to the shell view */
```
Comment (learn-more) [inline dll/win32/shell32/CNewMenu.cpp:446]: indenting changed?

File: dll/win32/shell32/CNewMenu.cpp [hunk 4]
```diff
@@ -406,38 +406,51 @@ CNewMenu::SHELLNEW_ITEM *CNewMenu::FindItemFromIdOffset(UINT IdOffset)
...
 HRESULT CNewMenu::SelectNewItem(LONG wEventId, UINT uFlags, LPWSTR pszName, BOOL bRename)
 {
-    CComPtr<IShellBrowser> lpSB;
+    CComPtr<IShellFolder> lpDesktopSF;
+    CComPtr<IShellFolder> lpCurrentSF;
     CComPtr<IShellView> lpSV;
     HRESULT hr = E_FAIL;
-    LPITEMIDLIST pidl;
-    PITEMID_CHILD pidlNewItem;
+    LPITEMIDLIST pidlFullNewItem = NULL;
```
Comment (learn-more) [inline dll/win32/shell32/CNewMenu.cpp:413]: If you use an `CComHeapPtr<ITEMIDLIST>` you do not need to free it manually

**Review by binarymaster — CHANGES_REQUESTED** [review]
Please address and respond to the review above.

File: dll/win32/shell32/CNewMenu.cpp [hunk 5]
```diff
@@ -406,38 +406,51 @@ CNewMenu::SHELLNEW_ITEM *CNewMenu::FindItemFromIdOffset(UINT IdOffset)
...
         return S_OK;
 
-    /* Get a pointer to the shell view */
-    hr = IUnknown_QueryService(m_pSite, SID_IFolderView, IID_PPV_ARG(IShellView, &lpSV));
-    if (FAILED_UNEXPECTEDLY(hr))
-        return S_OK;
-
     /* Attempt to get the pidl of the new item */
-    hr = SHILCreateFromPathW(pszName, &pidl, NULL);
+    hr = SHGetDesktopFolder(&lpDesktopSF);
```
[the remark this answers is not in the document]
Comment (bho3538) [reply dll/win32/shell32/CNewMenu.cpp:426]: image

It works fine if parent folder does not under 'Desktop'

'SHGetDesktopFolder' function will return 'root' of the shell namespace folder (not C:\Users\username\Desktop)

To get correct location folder object (ex Desktop->awefawef AND C:\Users->username->Desktop->awefawef are different pidl),
Query current folder pidl(m_pidlFolder) with IID_IShellFolder to root of the shell namespace folder.

After get correct parent folder object, try find created 'New Item's pidl with 'new item's name using 'ParseDisplayName'

Then call 'SHChangeNotify' to notify at shell 'New item' was created.

Shell will update new-item like folder or file.

**Comment by bho3538** [comment]
change code style and left comment

File: dll/win32/shell32/CNewMenu.cpp [hunk 6]
```diff
@@ -406,38 +406,53 @@ CNewMenu::SHELLNEW_ITEM *CNewMenu::FindItemFromIdOffset(UINT IdOffset)
...
+        {
+            pidlFullNewItem = ILCombine(m_pidlFolder, pidlNewItem);
+			
+            /* Notify the view object about the new item */
+            SHChangeNotify(wEventId, uFlags, (LPCVOID)pidlFullNewItem, NULL);
+            SHFree(pidlFullNewItem);
+        }
+    }
+	
+	/* Get a pointer to the shell view */
```
Comment (HBelusca) [inline dll/win32/shell32/CNewMenu.cpp:448]:
```suggestion
    /* Get a pointer to the shell view */
```

File: dll/win32/shell32/CNewMenu.cpp [hunk 7]
```diff
@@ -406,38 +406,51 @@ CNewMenu::SHELLNEW_ITEM *CNewMenu::FindItemFromIdOffset(UINT IdOffset)
...
-
     if (!m_pSite)
         return S_OK;
 
-    /* Get a pointer to the shell view */
-    hr = IUnknown_QueryService(m_pSite, SID_IFolderView, IID_PPV_ARG(IShellView, &lpSV));
-    if (FAILED_UNEXPECTEDLY(hr))
-        return S_OK;
-
     /* Attempt to get the pidl of the new item */
```
[the remark this answers is not in the document]
Comment (bho3538) [reply dll/win32/shell32/CNewMenu.cpp:425]: This code moved back after 'SHChangeNotify' with 'New Item' PIDL (not removed)

**Review by bho3538 — COMMENTED** [review]
left comment

**Comment by HBelusca** [comment]
Please also check the other unresolved comments.

File: dll/win32/shell32/CNewMenu.cpp (hunk 2 above)
Comment (bho3538) [reply dll/win32/shell32/CNewMenu.cpp:414]: I got to know something new. Do I need to change the code?

sorry for late reply. my reply was in pending mode.

**Pull request closed without merging**